src/app/qgssubstitutionlistwidget.cpp

```diff
@@ -0,0 +1,230 @@
+{
+  QList< QgsStringReplacement > result;
+  for ( int i = 0; i < mTableSubstitutions->rowCount(); ++i )
+  {
+    if ( !mTableSubstitutions->item( i, 0 ) )
+      continue;
+
+    if ( mTableSubstitutions->item( i, 0 )->text().isEmpty() )
+      continue;
+
+    QCheckBox* chkCaseSensitive = qobject_cast<QCheckBox*>( mTableSubstitutions->cellWidget( i, 2 ) );
+    QCheckBox* chkWholeWord = qobject_cast<QCheckBox*>( mTableSubstitutions->cellWidget( i, 3 ) );
+
+    QgsStringReplacement replacement( mTableSubstitutions->item( i, 0 )->text(),
+                                      mTableSubstitutions->item( i, 1 )->text(),
+                                      chkCaseSensitive->isChecked(),
+                                      chkWholeWord->isChecked() );
+    result << replacement;
+  }
+  return QgsStringReplacementCollection( result );
+}
+
+void QgsSubstitutionListWidget::on_mButtonAdd_clicked()
+{
+  addSubstitution( QgsStringReplacement( QString(), QString(), false, true ) );
+  mTableSubstitutions->setFocus();
+  mTableSubstitutions->setCurrentCell( mTableSubstitutions->rowCount() - 1, 0 );
+}
+
+void QgsSubstitutionListWidget::on_mButtonRemove_clicked()
+{
+  int currentRow = mTableSubstitutions->currentRow();
+  mTableSubstitutions->removeRow( currentRow );
+  tableChanged();
+}
+
+void QgsSubstitutionListWidget::tableChanged()
+{
+  emit substitutionsChanged( substitutions() );
+}
+
+void QgsSubstitutionListWidget::on_mButtonExport_clicked()
+{
+  QString fileName = QFileDialog::getSaveFileName( this, tr( "Save substitutions" ), QDir::homePath(),
+                     tr( "XML files (*.xml *.XML)" ) );
+  if ( fileName.isEmpty() )
+  {
+    return;
+  }
+
+  // ensure the user never ommited the extension from the file name
+  if ( !fileName.endsWith( ".xml", Qt::CaseInsensitive ) )
+  {
+    fileName += ".xml";
+  }
+
+  QDomDocument doc;
+  QDomElement root = doc.createElement( "substitutions" );
+  root.setAttribute( "version", "1.0" );
+  QgsStringReplacementCollection collection = substitutions();
+  collection.writeXml( root, doc );
+  doc.appendChild( root );
+
+  QFile file( fileName );
+  if ( !file.open( QIODevice::WriteOnly | QIODevice::Text ) )
+  {
+    QMessageBox::warning( nullptr, tr( "Export substitutions" ),
+                          tr( "Cannot write file %1:\n%2." ).arg( fileName, file.errorString() ),
+                          QMessageBox::Ok,
+                          QMessageBox::Ok );
+    return;
+  }
+
+  QTextStream out( &file );
+  doc.save( out, 4 );
+}
+
+void QgsSubstitutionListWidget::on_mButtonImport_clicked()
+{
+  QString fileName = QFileDialog::getOpenFileName( this, tr( "Load substitutions" ), QDir::homePath(),
+                     tr( "XML files (*.xml *.XML)" ) );
+  if ( fileName.isEmpty() )
+  {
+    return;
+  }
+
+  QFile file( fileName );
+  if ( !file.open( QIODevice::ReadOnly | QIODevice::Text ) )
+  {
+    QMessageBox::warning( nullptr, tr( "Import substitutions" ),
```

Comment:
In the future I think a generic way to handle error messages would be nice
but for now I can't think of a good solution.  It is possible to put the
message bar into widgets and just reuse it like that.  I have have
considered some magic to inject the messagebar into the top level widget
however didn't take it too far once I ran into issues.  One possible
solution is to just float the message bar in the active widget at the top
although that does mean you have to handle the resizing events, etc.

On Mon, Aug 29, 2016 at 9:05 PM, Matthias Kuhn <notifications@github.com>
wrote:

> In src/app/qgssubstitutionlistwidget.cpp
> <https://github.com/qgis/QGIS/pull/3438#discussion_r76589638>:
>
> > +  doc.save( out, 4 );
> > +}
> > +
> > +void QgsSubstitutionListWidget::on_mButtonImport_clicked()
> > +{
> > +  QString fileName = QFileDialog::getOpenFileName( this, tr( "Load substitutions" ), QDir::homePath(),
> > +                     tr( "XML files (*.xml *.XML)" ) );
> > +  if ( fileName.isEmpty() )
> > +  {
> > +    return;
> > +  }
> > +
> > +  QFile file( fileName );
> > +  if ( !file.open( QIODevice::ReadOnly | QIODevice::Text ) )
> > +  {
> > +    QMessageBox::warning( nullptr, tr( "Import substitutions" ),
>
> I agree that the message bar has also been abused a lot for out-of-context
> error messages. It's nice but not a solution to all.
> In this case here I don't mind a messagebox. Or if we want to go fancy a
> background and tooltip color for the line edit.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/3438/files/46fba7ce80039d5feb8445410ddf98e485f5951f#r76589638>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXS3HF_HqxmdCcWx-THsUBVSjPLakQRks5qkr0PgaJpZM4JvNYh>
> .
>